issue/142: ie8-ie9 support

oliverfoster opened the pull request:
#142 

### Code changes
* Reverted [`buildMultipartPost`](https://github.com/adlnet/xAPIWrapper/commit/09db62ddf12f1e7f1f5a4dfe4e537d920aed57fb) to the pre Uint8Array and Blob implementation to provide IE8-IE9 attachment support.
* Removed unused functions `ADL.stringToArrayBuffer`, `ADL.stringFromArrayBuffer` and `getByteLen`.
* Removed global declarations of various functions by keeping them inside a closure or replacing their global polyfill.
* Added `ADL.dateToISOString` as a replacement to `Date.prototype.toISOString` polyfill. [Ref](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toISOString#Polyfill)
* Indented `xapi-launch.js`.
* Added `ObjectKeys` where `Object.keys` is required but not supported. [Ref](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys#Polyfill)
* Removed `window` global use where unnecessary.
* Changed `ADL.xapiutil.getLang` to use `ADL.language` if defined.
* Applied [pull/144](https://github.com/adlnet/xAPIWrapper/pull/144).

### package.json
* Updated `doxstrap` dependency to allow docs to be built on later versions of node. [Ref](https://github.com/webpro/doxstrap/pull/7)
* Version bump from 1.10.x to 1.11 to show size of changes.

### Doc and testing changes
* Added `ADL.language` to allow override of environment language for testing.
* Added console support for IE8/IE9. [Ref](http://perrymitchell.net/article/ie8-javascript-console/)
* Made sure `CryptoJS` loads globally in NodeJS to allow testing.
* Used NodeJS url parser in NodeJS to support testing.
* Prevented `parseQueryString` from running in NodeJS to enable tests and stop attempts at accessing `window.location` in NodeJS.
* Forced browser examples and tests to run in non-compatibility mode in IE8 and IE9. [Ref](https://blogs.msdn.microsoft.com/askie/2009/03/23/understanding-compatibility-modes-in-internet-explorer-8/)
* Switched examples and tests to use jquery v1 instead of jquery v2 for IE8 support. [Ref](https://jquery.com/browser-support/#content)
* Changed module declaration of each file to self declare the ADL namespace globally in browser and NodeJS (for tests).
* Fixed bad preventDefault in `useWrapper.html`.
* Fixed `useWrapper.html` to correctly report returned value when using `XDomainRequest` in IE8.
* Added JSDoc comments throughout and added relevant files to documentation construction. Changed parameter names to ensure compliant creating of documentation.
* Removed mocha blanket test coverage plugin as unsupported in IE8.
* Removed assertion libraries should and expect and replaced with an IE8 compatible assert library.
* Changed all tests to use assert assertion library.
* Modified mocha.css to provide IE8 support on :before and :after. [Ref](https://developer.mozilla.org/en-US/docs/Web/CSS/::before)
* Moved statement testing mock to `test/mocks/ADL.XHR_request.js` from `test/my.mock.wrapper.js`.
* Moved tests from `test/tests` to `test` to support `$ mocha` command.
* Updated tests to work in both browser and node.

### Notes on testing
* IE8 and IE9 VMs can be fetched from Microsoft for testing. [Ref](https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/)
* You should run `useWrapper.html` from an HTTPS server as `XDomainRequest` is requesting from an HTTPS LRS and the cross domain requests over `XDomainRequest` must have the same protocol. [Ref](https://developer.mozilla.org/en-US/docs/Web/API/XDomainRequest#Security).
* XAMMP contains a preconfigured HTTPS webserver if you need one for testing. [Ref](https://www.apachefriends.org/index.html)

### Conclusion
* All tests working in browser and NodeJS.
* Documentation is working and is more exhaustive.
* Library works in IE8 and IE9.

vbhayden commented:
`Files Changed 31` makes me think this might be a little too big for a regular PR, but the added support is appreciated. 

Have been out sick, but will look this over when I get a chance.

oliverfoster commented:
I would say the same usually but my brain broke when I tried to break it into smaller pieces as a lot of it is interconnected.

FlorianTolk commented:
If you would be willing to make this break, it may make it easier for ADL to debug the changes locally before merging the two branches vs the one.

oliverfoster commented:
I'll endeavour to attend the call tomorrow and we can talk it through if that helps? I think I've worked out a few PRs I can do separately, I'd just like to agree them before I go forward. 

On reflection I'm not sure it'll work well giving you the tests without a working library or the library without working tests.

vbhayden commented:
I'm still thinking this should be pruned to just ie8/9 relevance.  At a glance, these aren't related to ie8/9:

## Code changes
- Removed unused functions ADL.stringToArrayBuffer, ADL.stringFromArrayBuffer and getByteLen.
- Removed global declarations of various functions by keeping them inside a closure or replacing their global polyfill.
- Indented xapi-launch.js.
- Removed window global use where unnecessary.
- Changed ADL.xapiutil.getLang to use ADL.language if defined.
- Applied pull/144.

## package.json
- Updated doxstrap dependency to allow docs to be built on later versions of node. Ref
- Doc and testing changes
- Added ADL.language to allow override of environment language for testing.
- Made sure CryptoJS loads globally in NodeJS to allow testing.
- Used NodeJS url parser in NodeJS to support testing.
- Prevented parseQueryString from running in NodeJS to enable tests and stop attempts at accessing window.location in NodeJS.
- Changed module declaration of each file to self declare the ADL namespace globally in browser and NodeJS (for tests).
- Fixed bad preventDefault in useWrapper.html.
- Added JSDoc comments throughout and added relevant files to documentation construction. Changed parameter names to ensure compliant creating of documentation.
- Changed all tests to use assert assertion library.
- Moved statement testing mock to test/mocks/ADL.XHR_request.js from test/my.mock.wrapper.js.
- Moved tests from test/tests to test to support $ mocha command.
- Updated tests to work in both browser and node.

oliverfoster commented:
Cool. That works for me. Except for the items below, which are required for ie8-9 compatibility as their behaviour is IE10+:

### Code changes
* Removed unused functions ADL.stringToArrayBuffer, ADL.stringFromArrayBuffer and getByteLen. (these become redundant once buildMultipartPost is reverted to not use Uint8Array and Blob)

### Doc and testing changes
* Changed all tests to use assert assertion library. (the existing assertion libraries don't work in ie8 and so need replacing)
* Fixed bad preventDefault in useWrapper.html (this stops the option/checkboxes from working correctly in ie8)

oliverfoster commented:
Is that proposal acceptable @vbhayden ?

vbhayden commented:
I'm fine with ensuring ie8/9 support, but would like the PR cleaned to include only changes relevant to that goal though.  

If you can split this into two, one for documentation / quality of life and another for ie8/9, then I'll get them reviewed.

oliverfoster commented:
It's a deal. Thanks for the direction. 👍

oliverfoster commented:
closed in favour of https://github.com/adlnet/xAPIWrapper/pull/147
